### PR TITLE
Fix #2178: testLineAgainstWater works incorrectly outside of game boundaries

### LINKED ISSUE #2178: testLineAgainstWater works incorrectly outside of game boundaries
**Describe the bug**
 `testLineAgainstWater` checks only for water at z-coordinate 0 outside of the -3000 : 3000 coords range. This behaviour was introduced along with multiple other fixes with this PR: https://github.com/multitheftauto/mtasa-blue/pull/568
 Some time afterwards, possiblity to change ocean water level outside of bounds was added for `setWaterLevel` function, thus rendering current behaviour incorrect.

**To reproduce**
Set ocean water level to something different than 0, use testLineAgainstWater to check for it, game still thinks the water is at z-coord 0

**Expected behaviour**
`testLineAgainstWater` should check for water at correct height

**Version**
current

**Additional context**
Fix is very simple, the offending line is: https://github.com/multitheftauto/mtasa-blue/blob/c58c3678f263d810f61eb5aecba48e6c4b4dce0c/Client/game_sa/CWaterManagerSA.cpp#L842
The fix is to change current logic to 
`vecStart.IntersectsSegmentPlane(rayDir, CVector(0, 0, 1), CVector(0, 0, <oceanWaterLevelHere>), &zeroPoint)`
where `<oceanWaterLevelHere>` is currently used z-coord of ocean water

### PR DESCRIPTION
**Pull request by Allerek**
Just added a check for water level outside Game Area, and commened out a check if both points are outside Game Area(why would we check for that lol).

Fixes https://github.com/multitheftauto/mtasa-blue/issues/2178

### CODE REVIEW

**Comment by forkerer** [comment]
The commented out check is an early out if both ends are on the same side of the world border, thus skipping all the further calculations, i think that the GetZonesIntersecting return zones at the world border even when both points are ourside of the -3000 : 3000 range
Red arrows show check that will be skipped, while the black one will go through
obraz

**Comment by forkerer** [comment]
I don't see a reason to remove that check, and i doubt you tested those cases or checked what happens in the code further down

**Comment by Allerek** [comment]
> I don't see a reason to remove that check, and i doubt you tested those cases or checked what happens in the code further down

Because thats what this PR is all about, if both of the ends will be outside area, it will automatically return false, we dont want that to happen, we want to be able to check if line is hitting the water in any case, and yeah - i tested that and it worked just fine.

Some users might want to get out of Game Area(for some reason, i'm not here to judge), and we should allow that.

**Comment by forkerer** [comment]
And they still can, the code you commented out doesn't change anything about that, it just skips unnecessary calculations later on. No water other than ocean water will ever be collided outside of -3000 : 3000 range because You can't create custom water outside that

**Comment by forkerer** [comment]
Check for the GetZonesIntersecting behaviour when you check line starting from (-3050, -3050, 10) to (3050, -3050, 10) when ocean water level is 0, it will do unnecessary checks around the world border if I'm not mistaken after You commented the code out

**Comment by Allerek** [comment]
> And they still can, the code you commented out doesn't change anything about that, it just skips unnecessary calculations later on. No water other than ocean water will ever be collided outside of -3000 : 3000 range because You can't create custom water outside that

Yeah, but the code returns false, it doesnt even check if the line really collides with  water, as far as i see it only  checks for X and Y.

-Are both points outside of map?
-If yes then check if they're on the same side of the map
-If so, just return false for some reason? < That doesnt make any sense, whats the reason to suggest theres no water in its way what soever?

**Comment by forkerer** [comment]
You literally check for ocean water collision few lines up, no other water will ever exist there, as it can't be created

**Comment by forkerer** [comment]
That's the only water you can encounter outside of game zone
obraz

**Comment by Allerek** [comment]
After thinking it out, you might be right, further code checks for waves as i see? Forgot the ocean water doesnt have waves at all, so that can make sense. Getting commented code back, leaving the higher fix tho.

**Comment by Allerek** [comment]
After all the checks are done, i guess we're ready to merge, as the change itself worked well in my tests.

File: Client/game_sa/CWaterManagerSA.cpp [hunk 1]
```diff
@@ -839,12 +839,16 @@ bool CWaterManagerSA::TestLineAgainstWater(const CVector& vecStart, const CVecto
 
     // Check if we're outside of map area.
     CVector zeroPoint;
-    if (vecStart.IntersectsSegmentPlane(rayDir, CVector(0, 0, 1), CVector(0, 0, 0), &zeroPoint) && IsPointOutsideOfGameArea(zeroPoint))
+    float   waterHeight;
+    CVector pvecUnknown;
```
Comment (Pirulax) [inline Client/game_sa/CWaterManagerSA.cpp]: Try passing nullptr for `vecUnknown`. It should work (judging from the source code..)

Comment (Allerek) [reply Client/game_sa/CWaterManagerSA.cpp]: I'll try that. Waiting for tests to pass.

Comment (Pirulax) [reply Client/game_sa/CWaterManagerSA.cpp]: I mean the CI will pass, just test it locally, see if everything is as expected.

**Comment by Allerek** [comment]
[watertest.zip](https://github.com/multitheftauto/mtasa-blue/files/6351383/watertest.zip)
Test resource.
image
(first number-current water level, second is endz for testLineAgainstWater)

**Comment by Allerek** [comment]
For some reason it returns true if water level is lower than 0, and we check at Z=0(or lower), i guess its due to
`GetWaterLevel(CVector(3000,3000,1), &waterHeight, true, nullptr);`
Where i check for Z level '1', i guess we could just set it to a very low number, but this aint the best solution i guess. Any ideas? @forkerer

**Comment by patrikjuvonen** [comment]
> For some reason it returns true if water level is lower than 0, and we check at Z=0(or lower), i guess its due to
> `GetWaterLevel(CVector(3000,3000,1), &waterHeight, true, nullptr);`
> Where i check for Z level '1', i guess we could just set it to a very low number, but this aint the best solution i guess. Any ideas? @forkerer

Did you check his suggested fix in the original issue #2178?

**Comment by Allerek** [comment]
> > For some reason it returns true if water level is lower than 0, and we check at Z=0(or lower), i guess its due to
> > `GetWaterLevel(CVector(3000,3000,1), &waterHeight, true, nullptr);`
> > Where i check for Z level '1', i guess we could just set it to a very low number, but this aint the best solution i guess. Any ideas? @forkerer
> 
> Did you check his suggested fix in the original issue #2178?

i got some help, i'll introduce it soon but for now i have some work to do

File: Client/game_sa/CWaterManagerSA.cpp [hunk 2]
```diff
@@ -838,10 +838,14 @@ bool CWaterManagerSA::TestLineAgainstWater(const CVector& vecStart, const CVecto
...
+
+    float waterHeight = *(float*)0x6E873F;
+
+    // This checks for intersections between Line and Plane(water is treated like a flat surface) outside a Game Area and returns a CVector with its coordinates if any were found.
+    if (vecStart.IntersectsSegmentPlane(rayDir, CVector(0, 0, 1), CVector(0, 0, waterHeight), &oceanWaterIntersectPoint) && IsPointOutsideOfGameArea(oceanWaterIntersectPoint))
     {
-        *vecCollision = zeroPoint;
+        *vecCollision = oceanWaterIntersectPoint;
         return true;
     }
```
Comment (Pirulax) [inline Client/game_sa/CWaterManagerSA.cpp:850]:
```suggestion
    // Check for intersection with ocean outside the game area (takes water level into account)
    // If a hit is detected, and it is outside, we early out, as custom water can't be created outside game boundaries
    {
        CVector intersection{};
        const float waterHeight = *static_cast<float*>(0x6E873F);
        if (vecStart.IntersectsSegmentPlane(rayDir, CVector(0, 0, 1), CVector(0, 0, waterHeight), &intersection))
        {
            if (IsPointOutsideOfGameArea(intersection))
            {
                *vecCollision = intersection;
                return true;
            }
        }
    }
```

**Review by Pirulax — CHANGES_REQUESTED** [review]
LGTM

File: Client/game_sa/CWaterManagerSA.cpp (hunk 2 above)
Comment (Pirulax) [inline Client/game_sa/CWaterManagerSA.cpp]:
```suggestion
    // Check for intersection with ocean outside the game area (takes water level into account)
    // If a hit is detected, and it is outside, we early out, as custom water can't be created outside game boundaries
    {
        CVector intersection{};
        const float waterHeight = *static_cast<float*>(0x6E873F);
        if (vecStart.IntersectsSegmentPlane(rayDir, CVector(0, 0, 1), CVector(0, 0, waterHeight), &intersection))
        {
            if (IsPointOutsideOfGameArea(intersection))
            {
                *vecCollision = intersection;
                return true;
            }
        }
    }
```

**Review by Pirulax — APPROVED** [review]
Well I hope you've tested it.

**Pull request merged**